src/kvstore/comm_tree.h

```diff
@@ -0,0 +1,500 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Copyright (c) 2018 by Contributors
+ */
+#ifndef MXNET_KVSTORE_COMM_TREE_H_
+#define MXNET_KVSTORE_COMM_TREE_H_
+#include <dmlc/omp.h>
+#include <string>
+#include <algorithm>
+#include <utility>
+#include <limits>
+#include <vector>
+#include <tuple>
+#include <thread>
+#include <map>
+#include "mxnet/ndarray.h"
+#include "gradient_compression.h"
+#include "../ndarray/ndarray_function.h"
+#include "../operator/tensor/sparse_retain-inl.h"
+#include "./kvstore_utils.h"
+#include "./gpu_topology.h"
+namespace mxnet {
+namespace kvstore {
+/**
+ * \brief an implementation of Comm that performs reduction on device
+ * directly using tree.
+ *
+ * It is faster if the total device-to-device bandwidths is larger than
+ * device-to-cpu, which is often true for 4 or 8 GPUs. But it uses more device
+ * memory.
+ */
+class CommDeviceTree : public CommDevice {
+ public:
+  CommDeviceTree() {
+    inited_ = false;
+    gpuarray_bound_ = dmlc::GetEnv("MXNET_KVSTORE_GPUARRAY_BOUND", 10000000);
+    backtrack_ = dmlc::GetEnv("MXNET_KVSTORE_BACKTRACK", 0);
+    link_usage_penalty_ = dmlc::GetEnv("MXNET_KVSTORE_LINK_USAGE_PENALTY", 0.7);
+  }
+
+  virtual ~CommDeviceTree() { }
+
+  void Init(int key, const NDArrayStorageType stype, const TShape& shape,
+            int dtype = mshadow::kFloat32) override {
+    tree_sorted_key_attrs_.emplace_back(key, shape, dtype);
+    sorted_key_attrs_.emplace_back(key, shape, dtype);
+  }
+
+  void InitBuffersAndComm(const std::vector<NDArray>& src) {
+    if (!inited_) {
+      for (const auto& a : src) {
+        devs_.push_back(a.ctx());
+      }
+      QueryTopology();
+      // Note: delayed allocation set to true, because we do not want to allocate
+      // both in TreeBufferEntry and BufferEntry, so we use a size_t to keep
+      // track of each key's shape within BufferEntry
+      // -this information is required for inherited Reduce- and
+      //  BroadcastRowSparse
+      InitMergeBuffer(devs_);
+      InitMergeBufferTree();
+      if (dmlc::GetEnv("MXNET_ENABLE_GPU_P2P", 1)) {
+        EnableP2P();
+      }
+    }
+  }
+
+  // src is sliced shape
+  // copy_buf not sliced
+  // merged not sliced
+  const NDArray& ReduceInner(int key, const std::vector<NDArray>& src, int root,
+                             int merged_row, int priority) {
+    std::vector<std::vector<NDArray>> reduce(devs_.size());
+
+    TreeBufferEntry& random_buf = tree_merge_buf_[0][key];
+    const NDArrayStorageType stype = random_buf.merged[0].storage_type();
+    std::vector<size_t>& topology = topology_[root];
+    NDArray buf_slice;
+
+    if (stype == kDefaultStorage) {
+      // Copy everything into buf.merged for each gpu
```

Comment:
We can avoid this copy, but it makes the code harder to read. Since the source and destination belong to the same GPU, the gain is minimal (10 samples/s i.e. 720 samples/s to 730 samples/s).

Reply:
I tested the throughput difference on a similar intra-GPU `CopyFromTo()`, by commenting out Line 306 in `comm_tree.h`. This ruins the correctness of the output, but it gives an idea of how much savings can be gotten. Testing on VGG-16 on an older commit, I got the following. Geomean difference is getting rid of one CopyFromTo makes it 2.2% faster.

```
v6: One more intra-GPU CopyFromTo than v7
fp32:
BS | v6   | v7
4  | 711  | 745
8  | 999  | 1035
16 | 1449 | 1478
32 | 1638 | 1672
64 | 1695 | 1739

fp16:
BS | v6   | v7
8  | 1552 | 1599
16 | 2127 | 2163
32 | 2916 | 2910
64 | 2720 | 2775
128| 2518 | 2532
```

Reply:
Thanks! 
Have you tried to use tools/bandwidth/measure.py to test the throughput?
According to my experiments, tree mode outperforms device mode on p3.x16.
However, on p2.x16, the performance of tree mode seems worse than device mode.
Could you double-check that?

Reply:
That makes to me, because p2.16x topology is fully connected using GPUDirect PCI-E. In fully connected networks, tree will never outperform device mode. 

The advantage on p3.16x is that device mode is forced to use 4 PCI-E (not GPUDirect, so must go through CPU/QPI), but tree is all NVLink, so it avoids CPU/QPI latency.

Reply:
Actually, p2.x16 has a worse communication topology compared to p3.x16.
For p2.x16, the topology is very special. The communication net is composed of 2 fully connected subnets: the first 9 gpus and the last 7 gpus. Unlike p3.x16, there are no connections between the 2 components.
In that case, I expect that tree mode should still outperform device mode. However, surprisingly the device mode is better. Actually, the performance is similar to local mode on p2.x16, which seems weird to me.